File: backend/src/controllers/direct-message.controller.ts
import type { Request, Response } from "express";
import { directMessageService } from "../services/direct-message.service.js";
import { ApiResponse } from "../utils/api-response.js";
import type { CreateMessageInput, ListConversationsQuery, ListMessagesQuery, StartConversationInput } from "../validators/direct-message.validator.js";

export class DirectMessageController {
  async startConversation(request: Request, response: Response) {
    const { receiverId } = request.body as StartConversationInput;
    const conversation = await directMessageService.startConversation(
      request.user!.id,
      receiverId
    );
    response.status(201).json(new ApiResponse(201, conversation, "Conversation started"));
  }

  async listConversations(request: Request, response: Response) {
    const conversations = await directMessageService.listConversations(
      request.user!.id,
      request.validated?.query as ListConversationsQuery
    );
    response.json(new ApiResponse(200, conversations, "Conversations retrieved"));
  }

  async getConversation(request: Request, response: Response) {
    const conversation = await directMessageService.getConversation(
      param(request, "conversationId"),
      request.user!.id
    );
    response.json(new ApiResponse(200, conversation, "Conversation retrieved"));
  }

  async getMessages(request: Request, response: Response) {
    const messages = await directMessageService.getMessages(
      param(request, "conversationId"),
      request.user!.id,
      request.validated?.query as ListMessagesQuery
    );
    response.json(new ApiResponse(200, messages, "Messages retrieved"));
  }

  async sendMessage(request: Request, response: Response) {
    const { message } = await directMessageService.sendMessage(
      param(request, "conversationId"),
      request.user!.id,
      request.body as CreateMessageInput,
      filesFrom(request)
    );
    response.status(201).json(new ApiResponse(201, message, "Message sent"));
  }

  async editMessage(request: Request, response: Response) {
    const message = await directMessageService.editMessage(
      param(request, "id"),
      request.user!.id,
      request.body.content
    );
    response.json(new ApiResponse(200, message, "Message edited"));
  }

  async deleteMessage(request: Request, response: Response) {
    const message = await directMessageService.deleteMessage(
      param(request, "id"),
      request.user!.id
    );
    response.json(new ApiResponse(200, message, "Message deleted"));
  }

  async deleteMessageForMe(request: Request, response: Response) {
    const result = await directMessageService.deleteMessageForMe(
      param(request, "id"),
      request.user!.id
    );
    response.json(new ApiResponse(200, result, "Message deleted for me"));
  }

  async deleteMessageForEveryone(request: Request, response: Response) {
    const message = await directMessageService.deleteMessageForEveryone(
      param(request, "id"),
      request.user!.id,
      request.user!.role === "ADMIN"
    );
    response.json(new ApiResponse(200, message, "Message deleted for everyone"));
  }

  async toggleReaction(request: Request, response: Response) {
    const reactions = await directMessageService.toggleReaction(
      param(request, "id"),
      request.user!.id,
      request.body.emoji,
      request.body.category
    );
    response.json(new ApiResponse(200, { reactions }, "Reaction updated"));
  }

  async toggleStar(request: Request, response: Response) {
    const result = await directMessageService.toggleStar(
      param(request, "id"),
      request.user!.id
    );
    response.json(new ApiResponse(200, result, result.isStarred ? "Message starred" : "Message unstarred"));
  }

  async togglePinMessage(request: Request, response: Response) {
    const result = await directMessageService.togglePinMessage(
      param(request, "id"),
      request.user!.id
    );
    response.json(new ApiResponse(200, result, result.isPinned ? "Message pinned" : "Message unpinned"));
  }

  async getStarredMessages(request: Request, response: Response) {
    const messages = await directMessageService.getStarredMessages(
      param(request, "conversationId"),
      request.user!.id
    );
    response.json(new ApiResponse(200, messages, "Starred messages retrieved"));
  }

  async getPinnedMessages(request: Request, response: Response) {
    const messages = await directMessageService.getPinnedMessages(
      param(request, "conversationId"),
      request.user!.id
    );
    response.json(new ApiResponse(200, messages, "Pinned messages retrieved"));
  }

  async forwardMessages(request: Request, response: Response) {
    const { messageIds, targetConversationIds } = request.body;
    const results = await directMessageService.forwardMessages(
      messageIds,
      targetConversationIds,
      request.user!.id
    );
    response.status(201).json(new ApiResponse(201, results, "Messages forwarded"));
  }

  async bulkDeleteForMe(request: Request, response: Response) {
    const result = await directMessageService.bulkDeleteForMe(
      param(request, "conversationId"),
      request.body.messageIds,
      request.user!.id
    );
    response.json(new ApiResponse(200, result, "Messages deleted for me"));
  }

  async bulkDeleteForEveryone(request: Request, response: Response) {
    const result = await directMessageService.bulkDeleteForEveryone(
      param(request, "conversationId"),
      request.body.messageIds,
      request.user!.id,
      request.user!.role === "ADMIN"
    );
    response.json(new ApiResponse(200, result, "Messages deleted for everyone"));
  }

  async bulkStar(request: Request, response: Response) {
    const result = await directMessageService.bulkStar(
      param(request, "conversationId"),
      request.body.messageIds,
      request.user!.id,
      request.body.star ?? true
    );
    response.json(new ApiResponse(200, result, "Messages star updated"));
  }

  async markAsRead(request: Request, response: Response) {
    await directMessageService.markAsRead(
      request.body.conversationId,
      request.user!.id
    );
    response.json(new ApiResponse(200, null, "Messages marked as read"));
  }

  async unreadCount(request: Request, response: Response) {
    const count = await directMessageService.getUnreadCount(request.user!.id);
    response.json(new ApiResponse(200, { count }, "Unread count retrieved"));
  }

  async togglePin(request: Request, response: Response) {
    const isPinned = await directMessageService.togglePin(
      param(request, "conversationId"),
      request.user!.id
    );
    response.json(new ApiResponse(200, { isPinned }, isPinned ? "Conversation pinned" : "Conversation unpinned"));
  }

  async toggleMute(request: Request, response: Response) {
    const isMuted = await directMessageService.toggleMute(
      param(request, "conversationId"),
      request.user!.id
    );
    response.json(new ApiResponse(200, { isMuted }, isMuted ? "Conversation muted" : "Conversation unmuted"));
  }

  async toggleArchive(request: Request, response: Response) {
    const isArchived = await directMessageService.toggleArchive(
      param(request, "conversationId"),
      request.user!.id
    );
    response.json(new ApiResponse(200, { isArchived }, isArchived ? "Conversation archived" : "Conversation unarchived"));
  }

  async markAsUnread(request: Request, response: Response) {
    await directMessageService.markAsUnread(
      param(request, "conversationId"),
      request.user!.id
    );
    response.json(new ApiResponse(200, null, "Conversation marked as unread"));
  }

  async markAsDelivered(request: Request, response: Response) {
    await directMessageService.markAsDelivered(
      param(request, "conversationId"),
      request.user!.id
    );
    response.json(new ApiResponse(200, null, "Messages marked as delivered"));
  }
}

const param = (request: Request, key: string): string => request.params[key] as string;

const filesFrom = (request: Request) =>
  Array.isArray(request.files) ? request.files : [];

export const directMessageController = new DirectMessageController();

File: backend-java/src/main/java/com/studyconnect/backend/controller/DirectMessageController.java
package com.studyconnect.backend.controller;

import com.studyconnect.backend.dto.ApiResponse;
import com.studyconnect.backend.dto.directmessage.EditDirectMessageRequest;
import com.studyconnect.backend.dto.directmessage.MarkAsReadRequest;
import com.studyconnect.backend.dto.directmessage.StartConversationRequest;
import com.studyconnect.backend.service.DirectMessageService;
import com.studyconnect.backend.util.SecurityUtil;
import jakarta.validation.Valid;
import java.util.List;
import org.springframework.security.core.Authentication;
import org.springframework.validation.annotation.Validated;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.ModelAttribute;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.multipart.MultipartFile;

@RestController
@Validated
@RequestMapping("/api/direct-messages")
public class DirectMessageController {

    private final DirectMessageService directMessageService;

    public DirectMessageController(DirectMessageService directMessageService) {
        this.directMessageService = directMessageService;
    }

    @PostMapping("/conversations")
    public ResponseEntity<ApiResponse<?>> startConversation(@Valid @RequestBody StartConversationRequest request, Authentication authentication) {
        return ResponseEntity.status(HttpStatus.CREATED).body(ApiResponse.success(
                directMessageService.startConversation(SecurityUtil.currentUserId(authentication), request.receiverId()),
                "Conversation started"
        ));
    }

    @GetMapping("/conversations")
    public ApiResponse<?> listConversations(
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "20") int limit,
            @RequestParam(required = false) String search,
            Authentication authentication) {
        return ApiResponse.success(
                directMessageService.listConversations(SecurityUtil.currentUserId(authentication), page, limit, search),
                "Conversations retrieved"
        );
    }

    @GetMapping("/conversations/{conversationId}")
    public ApiResponse<?> getConversation(@PathVariable String conversationId, Authentication authentication) {
        return ApiResponse.success(
                directMessageService.getConversation(conversationId, SecurityUtil.currentUserId(authentication)),
                "Conversation retrieved"
        );
    }

    @GetMapping("/conversations/{conversationId}/messages")
    public ApiResponse<?> getMessages(
            @PathVariable String conversationId,
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "30") int limit,
            @RequestParam(defaultValue = "latest") String order,
            @RequestParam(required = false) String search,
            Authentication authentication) {
        return ApiResponse.success(
                directMessageService.getMessages(conversationId, SecurityUtil.currentUserId(authentication), page, limit, order, search),
                "Messages retrieved"
        );
    }

    @PostMapping("/conversations/{conversationId}/messages")
    public ResponseEntity<ApiResponse<?>> sendMessage(
            @PathVariable String conversationId,
            @RequestParam(required = false, defaultValue = "") String content,
            @RequestParam(required = false) String replyTo,
            @RequestParam(name = "attachments", required = false) List<MultipartFile> attachments,
            Authentication authentication) {
        return ResponseEntity.status(HttpStatus.CREATED).body(ApiResponse.success(
                directMessageService.sendMessage(
                        conversationId,
                        SecurityUtil.currentUserId(authentication),
                        content,
                        replyTo,
                        attachments
                ),
                "Message sent"
        ));
    }

    @PutMapping("/messages/{messageId}")
    public ApiResponse<?> editMessage(
            @PathVariable String messageId,
            @Valid @RequestBody EditDirectMessageRequest request,
            Authentication authentication) {
        return ApiResponse.success(
                directMessageService.editMessage(messageId, SecurityUtil.currentUserId(authentication), request.content()),
                "Message edited"
        );
    }

    @DeleteMapping("/messages/{messageId}")
    public ApiResponse<?> deleteMessage(@PathVariable String messageId, Authentication authentication) {
        return ApiResponse.success(
                directMessageService.deleteMessage(messageId, SecurityUtil.currentUserId(authentication)),
                "Message deleted"
        );
    }

    @PostMapping("/messages/read")
    public ApiResponse<?> markAsRead(@Valid @RequestBody MarkAsReadRequest request, Authentication authentication) {
        directMessageService.markAsRead(request.conversationId(), SecurityUtil.currentUserId(authentication));
        return ApiResponse.success(null, "Messages marked as read");
    }

    @GetMapping("/conversations/unread")
    public ApiResponse<?> unreadCount(Authentication authentication) {
        return ApiResponse.success(
                java.util.Map.of("count", directMessageService.unreadCount(SecurityUtil.currentUserId(authentication))),
                "Unread count retrieved"
        );
    }
}
